src/lib/battery/battery_base.h

```diff
@@ -0,0 +1,183 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file battery_base.h
+ *
+ * Library calls for battery functionality.
+ *
+ * @author Julian Oes <julian@oes.ch>
+ * @author Timothy Scott <timothy@auterion.com>
+ */
+
+#pragma once
+
+#include <uORB/uORB.h>
+#include <uORB/topics/battery_status.h>
+#include <drivers/drv_hrt.h>
+#include <px4_module_params.h>
+#include <drivers/drv_adc.h>
+#include <board_config.h>
+#include <drivers/boards/common/board_common.h>
+
+/**
+ * BatteryBase is a base class for any type of battery.
+ *
+ * Each of the virtual _get_*() functions corresponds to a parameter used for calibration. However, depending
+ * on the type of battery, these values may not come from parameters.
+ *
+ * Most implementations of BatteryBase will also inherit ModuleParams, and most of the virtual function
+ * implementations will look something like:
+ * 		float _get_bat_v_empty() override {return _param_bat_v_empty().get();}
+ *
+ * For a full implementation, see src/modules/sensors/analog_battery.{cpp, h}
+ * For a minimal implementation, see src/modules/simulator/simulator.h
+ */
+class BatteryBase : ModuleParams
+{
+public:
+	BatteryBase();
+
+	/**
+	 * Reset all battery stats and report invalid/nothing.
+	 */
+	void reset();
+
+	/**
+	 * Get the battery cell count
+	 */
+	int cell_count() { return _get_bat_n_cells(); }
+
+	/**
+	 * Get the empty voltage per cell
+	 */
+	float empty_cell_voltage() { return _get_bat_v_empty(); }
+
+	/**
+	 * Get the full voltage per cell
+	 */
+	float full_cell_voltage() { return _get_bat_v_charged(); }
+
+	/**
+	 * Update current battery status message.
+	 *
+	 * @param voltage_raw Battery voltage read from ADC, in Volts
+	 * @param current_raw Voltage of current sense resistor, in Amps
+	 * @param timestamp Time at which the ADC was read (use hrt_absolute_time())
+	 * @param selected_source This battery is on the brick that the selected source for selected_source
+	 * @param priority: The brick number -1. The term priority refers to the Vn connection on the LTC4417
+	 * @param throttle_normalized Throttle of the vehicle, between 0 and 1
+	 * @param armed Arming state of the vehicle
+	 * @param shouldPublish If True, this function published a battery_status uORB message.
+	 */
+	void updateBatteryStatus(hrt_abstime timestamp, float voltage_v, float current_a, bool connected,
+				 bool selected_source, int priority, float throttle_normalized, bool armed, bool shouldPublish = true);
```

Comment:
Can we remove the default for `shouldPublish`? And call it `should_publish`. It's easy to get these bools wrong when there are many args, and if there are defaults it's even trickier.

Reply:
I spent too much time last semester in Python, where any function can have a hundred optional parameters!